### PR TITLE
Add preprocessing.GapEncoder

### LINKED ISSUE #1439: Online GapEncoder
[skrub](https://skrub-data.org/stable/index.html) is a wonderful new project related to scikit-learn. You can see Gaël Varoquaux present it [here](https://www.youtube.com/watch?v=I4Sksd93bZ4). They have a transformer called `GapEncoder`: it's a way to embed fuzzy strings. This could be really powerful online, say for classifying Tweets or Twitch messages, where typos are aplenty.

We already have a way to do online TD-IDF/count vectorization. But we don't have Gamma-Poisson matrix factorization. It is doable online though. Once we have it, we could assemble the two into a nice GapEncoder class. See paper [here](https://inria.hal.science/hal-02171256v4).

This is related to #1412. Indeed, maybe this works well without Gamma-Poisson matrix factorization. For instance, we could use `decomposition.LDA`, which we already have.

### PR DESCRIPTION
**Pull request by Fazel94**
Closes #1439.

This adds an online `GapEncoder` to `preprocessing`. It's the streaming version of skrub's `GapEncoder`: each string is turned into character n-gram counts and factorized into a few latent topics with a Gamma-Poisson model, learning one sample at a time. Fuzzy variants of the same string ("london", "London, UK", "Lomdon") share most of their n-grams, so they end up on the same topic — which is what makes it useful for messy categorical text like hand-typed city names or chat messages.

Following the discussion on the issue:

- The n-gram vocabulary grows online, the same way `preprocessing.LDA` grows its word vocabulary (unbounded for now, no `maximum_size_vocabulary` cap).
- Topics update through `A`/`B` accumulators with a `rho` forgetting factor.
- `transform_one` is read-only: it only touches n-grams already seen and never mutates the model.

There's a runnable example in the docstring (the fuzzy city-name case @MaxHalford asked about) that shows typo'd inputs landing on the right topic. It passes `check_estimator`, and I added behavioral tests covering the clustering, transform purity, vocabulary growth, seeding, and the `on`/`strip_accents`/`lowercase` options.

I'll follow up with a fuller worked example under `docs/examples` in a separate PR.

### CODE REVIEW

**Comment by MaxHalford** [comment]
Thank you! I am going on a break for a couple of weeks, so I will check your PRSs when I'm back.

**Comment by Fazel94** [comment]
I wish you a fun holiday. @MaxHalford

**Comment by MaxHalford** [comment]
I like it, it looks like a very good start! Could you provide some throughput benchmarks to get a sense of how fast/slow this is?

**Comment by Fazel94** [comment]
Thank you,  
I will send you the benchmarks at Wednesday.

**Comment by Fazel94** [comment]
## Benchmark

### Data
10 city names(London, Berlin, ...) corrupted (by deleting a letter, swaping two, doubling one) probably adding a suffix(", UK", "city", ...) and sometimes lowercased, eg "londo, uk", "London city".

4000/1000 train/test.

### Models
Skrub vs this implementation,  Both `n_components=10`.

### Speed and Quality
| | train | transform | nearest-neighbour accuracy |
|---|---|---|---|
| river GapEncoder (online, one pass) | 4,758 rows/s | 9,239 rows/s | 0.990 |
| skrub GapEncoder (batch, 5 iters, k-means++ init) | 4,761 rows/s | 16,786 rows/s | 0.997 |
| skrub GapEncoder (batch, 5 iters, random init) | 4,718 rows/s | 14,924 rows/s | 0.993 |


### Analysis
Same training speed, transform is half the speed due to using `transform_one` on each new example vs batching in Skrub.  

Setting `rho` to 0.999 make the accuracy 0.990 but a `rho` of 0.95 leads to 0.947 accuracy, It is due to `rho` in skrub in batch and in current implementation applied per example, that causes forgetting value 0.95 to be very harsh.

**Comment by Fazel94** [comment]
I can calculate `rho` to be rescaled to match Skrub in a given `batch_size`, I call it `half_life` it would be `self.decay = 0.5 ** (1 / half_life)` defaults to 1000(effectively `rho=0.993`), putting to `float('inf')` makes the model to never forget.

`rho` doesn't fit here since it is applied per-batch and here we process per example and thus  setting a value to similar Skrub leads to over-forgetting.

File: river/preprocessing/gap_encoder.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    training:
+
+    >>> for x in ["London, UK", "Lndon", "Paris, France", "Pariss"]:
+    ...     topics = enc.transform_one(x)
+    ...     print(x, max(topics, key=topics.get), {k: round(v, 3) for k, v in topics.items()})
+    London, UK 1 {0: 0.053, 1: 12.047}
+    Lndon 1 {0: 0.05, 1: 3.05}
+    Paris, France 0 {0: 16.547, 1: 0.053}
+    Pariss 0 {0: 4.549, 1: 0.051}
+
```
Comment (MaxHalford) [inline river/preprocessing/gap_encoder.py]: One last thing: could you add an example that puts a `GapEncoder` in a classification pipeline?

**Review by MaxHalford — APPROVED** [review]
Great stuff! It's great work, and a useful addition. Just asked for one thing in the docstring.

File: river/preprocessing/gap_encoder.py (hunk 1 above)
Comment (Fazel94) [reply river/preprocessing/gap_encoder.py]: Thank you for your kindness, will do that, next week.

**Comment by MaxHalford** [comment]
@Fazel94 I have an ask: could you add a benchmark to CodSpeed? I would like us to do that systematically from now on, when adding new estimators.

**Comment by Fazel94** [comment]
Of course, I'll be on it the next week.

**Comment by MaxHalford** [comment]
The failing unit test is flaky and unrelated, I'm ok merging. Good work!

**Comment by Fazel94** [comment]
Do you want me to take a look at the flaky unit test for another pr?

**Comment by MaxHalford** [comment]
I guess the next one who sees it can fix it :)
